beep/structure/base.py

```diff
@@ -0,0 +1,1373 @@
+            (BEEPDatapath): A child class of BEEPDatapath.
+
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_json_file(cls, filename):
+        """Load a structured run previously saved to file.
+
+        Loads a BEEPDatapath or (legacy) ProcessedCyclerRun structured object from json.
+
+        Can be used in combination with files serialized with BEEPDatapath.to_json_file.
+
+        Args:
+            filename (str, Pathlike): a json file from a structured run, serialzed with to_json_file.
+
+        Returns:
+            None
+        """
+        with open(filename, "r") as f:
+            d = json.load(f)
+
+        # Add this structured file path to the paths dict
+        paths = d.get("paths", {})
+        paths["structured"] = os.path.abspath(filename)
+        d["paths"] = paths
+
+        return cls.from_dict(d)
+
+    def to_json_file(self, filename):
+        """Save a BEEPDatapath to disk as a json.
+
+        Not named from_json to avoid conflict with MSONable.from_json(*)
+
+        Args:
+            filename (str, Pathlike): The filename to save the file to.
+
+        Returns:
+            None
+        """
+        with open(filename, "w") as f:
+            json.dump(self.as_dict(), f)
+
+    @StructuringDecorators.must_not_be_legacy
+    def as_dict(self):
+        """Serialize a BEEPDatapath as a dictionary.
+
+        Must not be loaded from legacy.
+
+        Returns:
+            (dict): corresponding to dictionary for serialization.
+
+        """
+
+        if not self.is_structured:
+            summary = None
+            cycles_interpolated = None
+            diagnostic_summary = None
+            diagnostic_interpolated = None
+        else:
+            summary = self.structured_summary.to_dict("list")
+            cycles_interpolated = self.structured_data.to_dict("list")
+            diagnostic_summary = self.diagnostic_summary.to_dict(
+                "list") if self.diagnostic_summary is not None else None
+            diagnostic_interpolated = self.diagnostic_data.to_dict("list") if self.diagnostic_data is not None else None
+
+        return {
+            "@module": self.__class__.__module__,
+            "@class": self.__class__.__name__,
+
+            # Vital things needed for BEEPDatapath
+            "raw_data": self.raw_data.to_dict("list"),
+            "metadata": self.metadata.raw,
+            "paths": self.paths,
+
+            # For backwards compatibility
+            # All this data is in the "metadata" key, this is just for redundancy
+            "barcode": self.metadata.barcode,
+            "protocol": self.metadata.protocol,
+            "channel_id": self.metadata.channel_id,
+
+            # Structured data, expensively obtained
+            "summary": summary,
+            "cycles_interpolated": cycles_interpolated,
+            "diagnostic_summary": diagnostic_summary,
+            "diagnostic_interpolated": diagnostic_interpolated
+        }
+
```

Comment:
Just to make sure I understand, both raw and interpolated data will be serialized together? Does this slow down the de-serialization? @patrickherring-TRI, would it be helpful to have an option to 'chunk' the deserialization to only load processed or raw data attributes, for instance? Maybe the performance difference is negligible, idk..

Reply:
I ask, cause quite often, we don't end up dealing with the raw time series data once it is structured. I imagine we can drop it after loading the object, but I was not sure if it slows things down to re-load it each time.

Reply:
Yeah that's a good observation.

The downside of saving `raw_data` is that the saved json are bigger. Depending on the file, 3-4x as large as json from PCR. After compression though, as suggested in #208, `BEEPDatapaths` _with_ raw_data are 2-5x smaller than they were with PCR (not including the raw data). Decompression and loading for the large files I tested (mostly just the test maccor one from AWS) is pretty fast, though not instant. 

The upside is that including the raw data makes the object complete. For example, you can structure a Datapath, save it to disk, load it, _independently restructure any of the most important attributes with different structuring parameters_, save it again, etc. I couldn't see an easy way of doing this when raw data is not saved; maybe the interpolated dfs could be reversed to raw data if needed, but I wasn't smart enough to figure this out. Restructuring files might become important if, for example, something is found wrong with the method for getting diagnostic summaries. Without saving the raw data, you'd have to reload and redo the entire structuring for all files. By including the raw data, you can just change the method, load the previously structured data, regenerate _only_ the diagnostic summaries, and save again to disk.

But to address whether this is actually going to cause a performance problem, I added #211.